[updater] `check()` fails on private repos due to wrong header

Hi,

I'm trying to implement a closed-beta updater feature for my app, and just ran into a problem caused by the `check()` function overwriting my manually written headers. 

----
# Background

Since I'm connecting to a private repo, I'm using a GitHub API endpoint instead of the direct URL to `latest.json`. And the API for [getting a release asset](https://docs.github.com/en/rest/releases/assets?apiVersion=2022-11-28#get-a-release-asset) is a bit funky. If the `Accept` header is set to `application/vnd.github+json` (which is considered "normal use" by the API, I believe), GitHub will return metadata about the asset - not the actual asset itself. So when using the API to download `latest.json` the `Accept` header needs to be set to `application/octet-stream` instead.

Exact description in the [GitHub REST API docs](https://docs.github.com/en/rest/releases/assets?apiVersion=2022-11-28#get-a-release-asset):

> To download the asset's binary content:
> - If within a browser, fetch the location specified in the browser_download_url key provided in the response.
> - Alternatively, **set the Accept header of the request to [application/octet-stream](https://docs.github.com/rest/using-the-rest-api/getting-started-with-the-rest-api#media-types).** The API will either redirect the client to the location, or stream it directly if possible. API clients should handle both a 200 or 302 response.

----
# The problem

I'm building an `Updater` object using the `UpdaterBuilder` so that I can supply the correct endpoint and necessary headers.
``` Rust
let updater = app
        .updater_builder()
        .endpoints(vec![url])?
        .headers(headers)
        .build()?;
```

When I later call `updater.check()`
``` Rust
if let Some(update) = updater.check().await? {
    // [...]
}
```

I get an error that updater failed to deserialize the update response:
`[2025-04-10][15:41:39][tauri_plugin_updater::updater][ERROR] failed to deserialize update response: unexpected character 'l' while parsing major version number`

The reason Updater can't parse the json file is because it's actually working on the json file containing metadata about the asset, and not on the `latest.json` itself (verified by setting log-level to TRACE and looking at the console output). I'm almost certain that this is caused by `check()` injecting the aforementioned `Accept` header [here](https://github.com/tauri-apps/plugins-workspace/blob/c698e7259452d8556c3533ed53dbda4f6c2ca5f4/plugins/updater/src/updater.rs#L348C9-L348C86) setting it to `application/json`. I assume there are two `Accept` headers in the outgoing request, but that the GitHub backend decides to go with the one injected by `check()`.

----
# Appendix: Updating from private repos

Since I've been wrangling with these headers for a while now, I thought I'd write down the process for accessing a private asset through the GitHub API, as I'm sure someone will come across this issue working on a similar problem.

1. Download metadata about the latest release from `https://api.github.com/repos/{owner}/{repo}/releases/latest` using the following headers:
``` Rust
use reqwest::header::{HeaderMap, ACCEPT, AUTHORIZATION, USER_AGENT}; 

let mut headers = HeaderMap::new();
headers.insert(ACCEPT, "application/vnd.github+json".parse()?);
headers.insert(USER_AGENT, "my-tauri-app".parse()?);
headers.insert(AUTHORIZATION, format!("Bearer {}", PRIVATE_REPO_TOKEN).parse()?);
```
2. Parse the metadata json and look for the asset containing your `latest.json`. Extract the `url` field from here and use it in the next step. It should look something like this: `"url": "https://api.github.com/repos/{owner}/{repo}/releases/assets/805105863"`
3. Downoad the asset using the `url` found in the previous step. This time you need a different `Access` header, like mentioned before:
``` Rust
use reqwest::header::{HeaderMap, ACCEPT, AUTHORIZATION, USER_AGENT}; 

let mut headers = HeaderMap::new();
headers.insert(ACCEPT, "application/octet-stream".parse()?);
headers.insert(USER_AGENT, "my-tauri-app".parse()?);
headers.insert(AUTHORIZATION, format!("Bearer {}", PRIVATE_REPO_TOKEN).parse()?);
```
4. GitHub also recommends adding a header to specify the API Version you are integrating against. Everything works fine without this header at the time being, so I didn't insert it in the steps above for readability. But here it is anyway:
``` Rust
use reqwest::header::{HeaderMap, HeaderName, ACCEPT, AUTHORIZATION, USER_AGENT}; 

// Optional header that is recommended by GitHub
let hdr = HeaderName::from_lowercase(b"x-github-api-version")?;
headers.insert(hdr, "2022-11-28".parse()?);
```

--- a/.changes/updater-override-header.md
+++ b/.changes/updater-override-header.md
@@ -1,0 +1,6 @@
+---
+updater: patch
+updater-js: patch
+---
+
+Fix `check` and `download` overrides the `accept` header

--- a/plugins/updater/src/updater.rs
+++ b/plugins/updater/src/updater.rs
@@ -17,7 +17,7 @@ use std::ffi::OsStr;
 
 use base64::Engine;
 use futures_util::StreamExt;
-use http::HeaderName;
+use http::{header::ACCEPT, HeaderName};
 use minisign_verify::{PublicKey, Signature};
 use percent_encoding::{AsciiSet, CONTROLS};
 use reqwest::{
@@ -345,7 +345,9 @@ impl Updater {
     pub async fn check(&self) -> Result<Option<Update>> {
         // we want JSON only
         let mut headers = self.headers.clone();
-        headers.insert("Accept", HeaderValue::from_str("application/json").unwrap());
+        if !headers.contains_key(ACCEPT) {
+            headers.insert(ACCEPT, HeaderValue::from_static("application/json"));
+        }
 
         // Set SSL certs for linux if they aren't available.
         #[cfg(target_os = "linux")]
@@ -549,10 +551,9 @@ impl Update {
     ) -> Result<Vec<u8>> {
         // set our headers
         let mut headers = self.headers.clone();
-        headers.insert(
-            "Accept",
-            HeaderValue::from_str("application/octet-stream").unwrap(),
-        );
+        if !headers.contains_key(ACCEPT) {
+            headers.insert(ACCEPT, HeaderValue::from_static("application/octet-stream"));
+        }
 
         let mut request = ClientBuilder::new().user_agent(UPDATER_USER_AGENT);
         if let Some(timeout) = self.timeout {

